app/models/main.py

```diff
@@ -1832,6 +1848,307 @@ def serialize(self):
+                )),
+            ),
+            # if not awarded, we shouldn't have direct_award_search_id or direct_award_archived_service_id set
+            else_=sql_and(c(direct_award_search_id.is_(None)), c(direct_award_archived_service_id.is_(None))),
+        ), name="ck_process_outcomes_direct_award_keys_nullable"),
+        db.CheckConstraint(sql_case(
+            (
+                # if awarded, the "null-ness" of brief_response_id and brief_id must be the same
+                (result == "awarded", c(brief_response_id.is_(None)) == c(brief_id.is_(None)),),
+            ),
+            # if not awarded, brief_response_id should not be set
+            else_=c(brief_response_id.is_(None)),
+        ), name="ck_process_outcomes_brief_keys_nullable"),
+        # the "null-ness" of direct_award_project_id and brief_id must *never* be the same
+        db.CheckConstraint(
+            (c(direct_award_project_id.is_(None)) != c(brief_id.is_(None))),
+            name="ck_process_outcomes_either_brief_xor_direct_award",
+        ),
+        # only one completed award can point at a DirectAwardProject at once
+        db.Index(
+            "idx_process_outcomes_completed_direct_award_project_unique",
+            direct_award_project_id,
+            postgresql_where=(completed_at.isnot(None)),
+            unique=True,
+        ),
+        # only one completed award can point at a Brief at once
+        db.Index(
+            "idx_process_outcomes_completed_brief_unique",
+            brief_id,
+            postgresql_where=(completed_at.isnot(None)),
+            unique=True,
+        ),
+        # direct_award_archived_service_id must actually be a result of the given DirectAwardSearch
+        db.ForeignKeyConstraint(
+            (direct_award_archived_service_id, direct_award_search_id,),
+            ("direct_award_search_result_entries.archived_service_id", "direct_award_search_result_entries.search_id",),
+            name="fk_process_outcomes_da_service_id_da_search_id",
+            deferrable=True,
+            initially="DEFERRED",
+        ),
+        # direct_award_search_id must actually belong the given DirectAwardProject
+        db.ForeignKeyConstraint(
+            (direct_award_search_id, direct_award_project_id,),
+            ("direct_award_searches.id", "direct_award_searches.project_id",),
+            name="fk_process_outcomes_da_search_id_da_project_id",
+            deferrable=True,
+            initially="DEFERRED",
+        ),
+        # brief_response_id must actually belong to the given Brief
+        db.ForeignKeyConstraint(
+            (brief_response_id, brief_id,),
+            ("brief_responses.id", "brief_responses.brief_id",),
+            name="fk_process_outcomes_brief_response_id_brief_id",
+            deferrable=True,
+            initially="DEFERRED",
+        ),
+    )
+
+    # let's not leave our de-uglification wrapper hanging around as a method
+    del c
```

Comment:
Because otherwise it would give all process outcome instances this weird unintentional `c` method which was never meant to be used outside this block. Trying not to leak a reference to something I don't need to. Encapsulation.

(actually come to think I haven't *checked* this actually does what I expect...)

Reply:
So what you're worried about is polluting the namespace. In this case we have the opportunity to just not.
I understand the point. I totally get it. But defining, then deleting, this weird single letter attribute _at class definition_ is not a good trade off for the sake of a couple of newlines in code. It's just going to confuse people. It's smart and neat but it's also confusing for people not familiar with how python classes are constructed or not familiar with the _incredibly rarely used_ `del` keyword.

Just the fact that we're having this conversation proves that. The newlines might not be as pretty but their ugliness wouldn't have started a conversation.

Reply:
It's not "a few newlines", it's **significantly** bulking out most of the sql expressions used in this block with scary content that is actually _meaningless_, making them look much weirder than they actually are. If people don't know about the `del` keyword or how python classes are constructed, then hopefully they'll learn something - this isn't kindergarten.

Reply:
See my previous comment - there's a TODO about why it can go away, so let's not fret about it - the intention here is clear enough I think. Even a newbie will appreciate that something that is `del`eted is going to go away, and if they didn't understand much about why it was there then maybe they don't need to - there are references to the issue above anyway.